DEVOPS-203 - Update GitHub actions building appropriate php versions

ikhalid08 opened the pull request:
**Description:**
Once a new version has been pushed to the main branch, this PR will go through each PHP version and identify if it needs to be updated. If a change is detected, it will build and push the new image. It will also update the file in the repository so that it can be used for the future version check comparison.

grantlucas commented on .github/workflows/build-v80.yml:
```diff
@@ -0,0 +1,85 @@
+      # Does the Image exist
+      - name: Does the image exist?
+        run: |
+          # run command and capture stderr
```
Is this fully needed or can we just rely on the failure in the normal pull that doesn't write to a file?

ikhalid08 replied:
@grantlucas Turns out this is fully needed. 

Here is the action run where I do a regular pull:
https://github.com/hostpapa/nginx-fpm-alpine/actions/runs/6190798961/job/16807805314
As you can see it doesn't run the subsequent build command.

Here is the action run where the above step is used:
https://github.com/hostpapa/nginx-fpm-alpine/actions/runs/6190809189/job/16807837730
As you can see, it actually gets to building the image which is what we expect it to do.

grantlucas commented on .github/workflows/build-v80.yml:
```diff
@@ -0,0 +1,85 @@
+        run: |
+          # run command and capture stderr
+          docker pull ${{ env.REGISTRY }}/${{ github.event.repository.owner.login }}/${{ github.event.repository.name }}:${{ env.CURRENT_VERSION }} 2> errors.txt || true
+          echo ::set-output name=errors::`cat errors.txt`
```
When looking at the build details I saw [this warning about `set-output` being deprecated and removed soon](https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/)

grantlucas reviewed (changes requested):
@ikhalid08 should i be seeing the now tagged images in GHCR from the testing? In theory, I'd expect to see a run where it doesn't find the image, so it builds, tags, and pushes it. Then another run where it doesn't see any changes and doesn't build. to re-test, one can just delete the tag from GHCR to re-run. but I'm not seeing any instructions for how this was tested, should be tested, or verified.

ikhalid08 commented on .github/workflows/build-v80.yml:
```diff
@@ -0,0 +1,85 @@
+
+on:
+  push:
+    branches: [ main ]
```
@grantlucas I will be doing that after the changes are merged into the main so it can build the initial versions. Otherwise, the moment it merges, it will not fire off any builds since nothing is getting changed in those paths.

Hence, I need to merge this PR. Once it builds the initial images and pushes them to the registry, I will then add the path as well in a follow up PR.

grantlucas commented on .github/workflows/build-v80.yml:
```diff
@@ -0,0 +1,83 @@
+          labels: ${{ steps.meta.outputs.labels }}
+          file: ./8.0/alpine3.16/fpm-nginx/Dockerfile
+          build-args: |
+            ENVIRONMENT=development
```
What impact does this have on the image? I don't see it used anywhere

ikhalid08 commented on .github/workflows/build-v80.yml:
```diff
@@ -0,0 +1,92 @@
+
+      # Extract metadata (tags, labels) for Docker
+      # https://github.com/docker/metadata-action
+      - name: Extract Docker metadata
```
@grantlucas This is the step that I was missing. Upon studying this further, I realized it was missing this as per this article:
https://docs.github.com/en/actions/publishing-packages/publishing-docker-images

ikhalid08 replied:
Specially the image name

ikhalid08 replied:
I am going to pull in these changes and work on it to get the building of the image to pass and push the image.

grantlucas reviewed (changes requested):
The extract and sign steps should be based on the failure of the detection of the image as well. Extract docker, build, and sign should all only fire _if_ the image wasn't originally detected and thus needed building

ikhalid08 commented:
> The extract and sign steps should be based on the failure of the detection of the image as well. Extract docker, build, and sign should all only fire _if_ the image wasn't originally detected and thus needed building

@grantlucas Updated all three to only fire if the image doesn't get detected.

grantlucas reviewed (changes requested):
Looks like it's working well. I deleted the 8.1 tag manually and re-ran [this job](https://github.com/hostpapa/nginx-fpm-alpine/actions/runs/6203023056/job/16887884659) to validate that it went down the different path where it now had to build and push the image to verify it ended with a ✅ on the pipeline in the end. All worked well.

**One final change:**

The name of the job is `docker` which doesn't really mean anything. It's visible like this when in certain action views etc. For the verify template workflow it looks like you had defined better human `name:` options for each job. That'd be good standard practice going forwards. Make sure job key and names describe what they're doing. This should be something to do with building the image etc.

ikhalid08 commented:
> Looks like it's working well. I deleted the 8.1 tag manually and re-ran [this job](https://github.com/hostpapa/nginx-fpm-alpine/actions/runs/6203023056/job/16887884659) to validate that it went down the different path where it now had to build and push the image to verify it ended with a ✅ on the pipeline in the end. All worked well.
> 
> **One final change:**
> 
> The name of the job is `docker` which doesn't really mean anything. It's visible like this when in certain action views etc. For the verify template workflow it looks like you had defined better human `name:` options for each job. That'd be good standard practice going forwards. Make sure job key and names describe what they're doing. This should be something to do with building the image etc.

@grantlucas Good point on the naming. I have gone ahead and updated the job names for all three builds. It now reflects what is being done in the job itself.
